virocon/distributions.py

```diff
@@ -1494,3 +1495,143 @@ def _fit_mle(self, sample):
+        for key, arg in kwargs.items():
+            if key in self._param_names:
+                setattr(self, key, arg)
+            elif key.startswith("f_") and key[2:] in self._param_names:
+                setattr(self, key, arg)
+                setattr(self, key[2:], arg)
+            else:
+                raise TypeError(
+                    f"{self.__class__.__name__} got an unexpected keyword argument '{key}'"
+                )
+
+    #  https://stackoverflow.com/a/53640468
+    @staticmethod
+    def _list_scipy_parameters(distribution):
+        """List parameters for scipy.stats.distribution.
+        # Arguments
+            distribution: a string or scipy.stats distribution object.
+        # Returns
+            A list of distribution parameter strings.
+        """
+        if isinstance(distribution, str):
+            distribution = getattr(sts, distribution)
+        if distribution.shapes:
+            parameters = [name.strip() for name in distribution.shapes.split(",")]
+        else:
+            parameters = []
+        # assert distribution is continous
+        parameters += ["loc", "scale"]
+        return parameters
+
+    def _set_default_parameter_values(self):
+        for par_name in self._param_names:
+            if par_name == "loc":
+                setattr(self, par_name, 0)
+            else:
+                setattr(self, par_name, 1)
+            setattr(self, f"f_{par_name}", None)
+
+    @property
+    def parameters(self):
+        return {par_name: getattr(self, par_name) for par_name in self._param_names}
+
+    def _get_scipy_parameters(self, *args, **kwargs):
+        args_with_default = list(self.parameters.values())
+        for i, arg in enumerate(args):
+            if arg is not None:
+                args_with_default[i] = arg
+
+        for key, arg in kwargs.items():
+            try:
+                idx = self._param_names.index(key)
+            except ValueError:
+                # unknown parameter name
+                assert False
+            args_with_default[idx] = arg
+
+        return args_with_default
+
+    def cdf(self, x, *args, **kwargs):
+        scipy_par = self._get_scipy_parameters(*args, **kwargs)
+        return self.scipy_dist.cdf(x, *scipy_par)
+
+    def icdf(self, prob, *args, **kwargs):
+        scipy_par = self._get_scipy_parameters(*args, **kwargs)
+        return self.scipy_dist.ppf(prob, *scipy_par)
+
+    def pdf(self, x, *args, **kwargs):
+        scipy_par = self._get_scipy_parameters(*args, **kwargs)
+        return self.scipy_dist.pdf(x, *scipy_par)
+
+    def draw_sample(self, n, *args, random_state=None, **kwargs):
+        scipy_par = self._get_scipy_parameters(*args, **kwargs)
+        rvs_size = self._get_rvs_size(n, scipy_par)
+        return self.scipy_dist.rvs(*scipy_par, size=rvs_size, random_state=random_state)
+
+    def _fit_mle(self, sample):
+        p0 = [v for k, v in self.parameters.items() if k != "loc" and k != "scale"]
```

Comment:
Could you add a comment to explain this line?

I guess it produces a list of all distribution parameters which are not named "loc" and "scale"?

So the result for a 4-parameter distribution that has loc and scale parameters might be [0.0 1.0] (if 0 and 1 are the default values? This is a bit of a guess, I don't understand it.)

Reply:
Done.